### PR TITLE
[Enhancement] [plsql] support drop store procedure (#30966)

### LINKED ISSUE #30966: [Enhancement] [plsql] support drop store procedure
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

_No response_

### Solution

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by Vallishp**
## Proposed changes

Issue Number: close #30966

Support Drop stored procedure.

Unit test report.

mysql> delimiter //
mysql> CREATE OR REPLACE PROCEDURE  procedure_select( IN id_n INT)
    -> BEGIN
    ->          select * from plsql_tbl  where id = id_n;
    -> END
    -> //
Query OK, 0 rows affected (0.05 sec)

mysql> call procedure_select(111)
    -> ;//
+------+----------+
| id   | name     |
+------+----------+
|  111 | plsql111 |
+------+----------+
1 row in set (0.29 sec)

Query OK, 0 rows affected (0.29 sec)

mysql> drop procedure procedure_select;//
Query OK, 0 rows affected (0.03 sec)

mysql> call procedure_select(111) ;//
ERROR 1064 (HY000): plsql exec error, Ln:1 identifier 'procedure_select' must be declared.
mysql> 

Meta info before and after the drop:

after create the dump of meta from FE.

  "nameToStoredProcedures": [
    [
      {
        "name": "PROCEDURE_SELECT",
        "catalogName": "",
        "dbName": ""
      },
      {
        "name": "PROCEDURE_SELECT",
        "catalogName": "",
        "dbName": "",
        "ownerName": "root",
        "source": "CREATE OR REPLACE PROCEDURE  procedure_select( IN id_n INT)\nBEGIN\n         select * from plsql_tbl  where id \u003d id_n;\nEND"
      }
    ]
  ],
  "nameToPackages": {}
}
y"header

after drop

:{
  "nameToStoredProcedures": {},
  "nameToPackages": {}
}
y"header

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by morrySnow** [comment]
could u use text instead of pic in description? text is search engine and git commit msg friendly

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4 [hunk 1]
```diff
@@ -37,6 +37,7 @@ statement
     : statementBase # statementBaseAlias
     | CALL name=multipartIdentifier LEFT_PAREN (expression (COMMA expression)*)? RIGHT_PAREN #callProcedure
     | (ALTER | CREATE (OR REPLACE)? | REPLACE) (PROCEDURE | PROC) name=multipartIdentifier LEFT_PAREN .*? RIGHT_PAREN .*? #createProcedure
+    | (DROP) (PROCEDURE | PROC) name=multipartIdentifier #dropProcedure
```
Comment (morrySnow) [inline fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4:40]:
```suggestion
    | DROP (PROCEDURE | PROC) name=multipartIdentifier #dropProcedure
```

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/PLParser.g4 [hunk 2]
```diff
@@ -324,11 +325,15 @@ package_body_item :
...
     | create_function_stmt
     | create_procedure_stmt
+    | drop_procedure_stmt
     ;
 
 create_procedure_stmt :
       (ALTER | CREATE (OR REPLACE)? | REPLACE) (PROCEDURE | PROC) multipartIdentifier create_routine_params? create_routine_options? (AS | IS)? declare_block_inplace? label_stmt? procedure_block (ident_pl SEMICOLON)?
     ;
+drop_procedure_stmt:
+      (DROP  (PROCEDURE | PROC) multipartIdentifier create_routine_params? create_routine_options?)
```
Comment (morrySnow) [inline fe/fe-core/src/main/antlr4/org/apache/doris/nereids/PLParser.g4:335]: why syntax not same with drop proc in DorisParser.g4?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 3]
```diff
@@ -3286,4 +3288,15 @@ public LogicalPlan visitCreateProcedure(CreateProcedureContext ctx) {
...
+
+    @Override
+    public LogicalPlan visitDropProcedure(DropProcedureContext ctx) {
+        List<String> nameParts = visitMultipartIdentifier(ctx.name);
+        FuncNameInfo procedureName = new FuncNameInfo(nameParts);
+        return ParserUtils.withOrigin(ctx, () -> {
+            LogicalPlan dropProcedurePlan;
+            dropProcedurePlan = new DropProcedureCommand(procedureName, getOriginSql(ctx));
+            return dropProcedurePlan;
+        });
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:3300]:
```suggestion
        return ParserUtils.withOrigin(ctx, () -> DropProcedureCommand(procedureName, getOriginSql(ctx)));
```

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/PLParser.g4 (hunk 2 above)
Comment (Vallishp) [reply fe/fe-core/src/main/antlr4/org/apache/doris/nereids/PLParser.g4:335]: from stored procedure can be created using overloading with different parameter. so drop procedure kept similar to stored create procedure for plsql.

**Comment by Vallishp** [comment]
> could u use text instead of pic in description? text is search engine and git commit msg friendly

thanks for your comments. I will update it soon

**Comment by Vallishp** [comment]
> could u use text instead of pic in description? text is search engine and git commit msg friendly

update. thank you

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/PLParser.g4 (hunk 2 above)
Comment (xinyiZzz) [reply fe/fe-core/src/main/antlr4/org/apache/doris/nereids/PLParser.g4:335]: `create_routine_params` and `create_routine_options` are not required,
refer to /hive/hplsql/src/main/antlr4/org/apache/hive/hplsql/Hplsql.g4,
image

we keep the same grammar as hplsql, such as:
`DROP PROCEDURE (IF EXISTS)? name=multipartIdentifier`
or
`DROP (PROCEDURE | PROC) (IF EXISTS)? name=multipartIdentifier`

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4 [hunk 4]
```diff
@@ -37,6 +37,7 @@ statement
     : statementBase # statementBaseAlias
     | CALL name=multipartIdentifier LEFT_PAREN (expression (COMMA expression)*)? RIGHT_PAREN #callProcedure
     | (ALTER | CREATE (OR REPLACE)? | REPLACE) (PROCEDURE | PROC) name=multipartIdentifier LEFT_PAREN .*? RIGHT_PAREN .*? #createProcedure
+    | DROP (PROCEDURE | PROC) name=multipartIdentifier #dropProcedure
```
Comment (xinyiZzz) [inline fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4:40]: Same as `PLParser.g4` `drop_procedure_stmt`

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/PLParser.g4 [hunk 5]
```diff
@@ -324,11 +325,15 @@ package_body_item :
       declare_stmt_item
     | create_function_stmt
     | create_procedure_stmt
+    | drop_procedure_stmt
```
Comment (xinyiZzz) [inline fe/fe-core/src/main/antlr4/org/apache/doris/nereids/PLParser.g4:328]: In /hive/hplsql/src/main/antlr4/org/apache/hive/hplsql/Hplsql.g4
image

no `drop_procedure_stmt`, but I don't know why.

**Comment by xinyiZzz** [comment]
@Vallishp Doris plsql is modified based on hive hplsql, can continue to refer to hplsql grammar. If it conflicts with oracle, it will be compatible with oracle first.

File: fe/fe-core/src/main/java/org/apache/doris/plsql/functions/DorisFunctionRegistry.java [hunk 6]
```diff
@@ -207,6 +208,23 @@ private void saveInCache(String name, ParserRuleContext procCtx) {
...
+    @Override
+    public void removeUserProcedure(Drop_procedure_stmtContext ctx) {
+        FuncNameInfo procedureName = new FuncNameInfo(
+                exec.logicalPlanBuilder.visitMultipartIdentifier(ctx.multipartIdentifier()));
+        if (builtinFunctions.exists(procedureName.toString())) {
+            exec.info(ctx, procedureName.toString() + " is a built-in function which cannot be removed.");
+            return;
+        }
+        trace(ctx, "DROP PROCEDURE " + procedureName.toString());
+        removeInCache(procedureName.toString(), ctx);
```
Comment (xinyiZzz) [inline fe/fe-core/src/main/java/org/apache/doris/plsql/functions/DorisFunctionRegistry.java:220]: Why not use `DorisFunctionRegistry removeCached`

File: fe/fe-core/src/main/java/org/apache/doris/plsql/functions/InMemoryFunctionRegistry.java [hunk 7]
```diff
@@ -251,6 +252,20 @@ public void addUserProcedure(Create_procedure_stmtContext ctx) {
         procMap.put(procedureName.toString(), ctx);
     }
 
+    @Override
+    public void removeUserProcedure(Drop_procedure_stmtContext ctx) {
+        FuncNameInfo procedureName = new FuncNameInfo(
```
Comment (xinyiZzz) [inline fe/fe-core/src/main/java/org/apache/doris/plsql/functions/InMemoryFunctionRegistry.java:257]:
```
FuncNameInfo procedureName = new FuncNameInfo(
                exec.logicalPlanBuilder.visitMultipartIdentifier(ctx.multipartIdentifier()));
        if (builtinFunctions.exists(procedureName.toString())) {
            exec.info(ctx, procedureName.toString() + " is a built-in function which cannot be removed.");
            return;
        }
        if (trace) {
            trace(ctx, "DROP PROCEDURE " + procedureName.toString());
        }
```
Commonly used in `DorisFunctionRegistry.removeUserProcedure` and `InMemoryFunctionRegistry.removeUserProcedure`.

Consider moving to `visitDrop_procedure_stmt`

File: fe/fe-core/src/main/java/org/apache/doris/plsql/functions/InMemoryFunctionRegistry.java [hunk 8]
```diff
@@ -251,6 +252,20 @@ public void addUserProcedure(Create_procedure_stmtContext ctx) {
...
+                exec.logicalPlanBuilder.visitMultipartIdentifier(ctx.multipartIdentifier()));
+        if (builtinFunctions.exists(procedureName.toString())) {
+            exec.info(ctx, procedureName.toString() + " is a built-in function which cannot be removed.");
+            return;
+        }
+        if (trace) {
+            trace(ctx, "DROP PROCEDURE " + procedureName.toString());
+        }
+        procMap.remove(procedureName.toString());
```
Comment (xinyiZzz) [inline fe/fe-core/src/main/java/org/apache/doris/plsql/functions/InMemoryFunctionRegistry.java:266]: Consider using `InMemoryFunctionRegistry.remove`.

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/PLParser.g4 (hunk 5 above)
Comment (Vallishp) [reply fe/fe-core/src/main/antlr4/org/apache/doris/nereids/PLParser.g4:328]: I checked hplsql.g4, I think it's implemented as drop_stmt. do you think we should follow same ?  I had followed createprocedure method.  

what is your WeChat id? can I add you there?

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/PLParser.g4 (hunk 2 above)
Comment (xinyiZzz) [inline fe/fe-core/src/main/antlr4/org/apache/doris/nereids/PLParser.g4:335]: not need `create_routine_params` and `create_routine_options`, refer to /hive/hplsql/src/main/antlr4/org/apache/hive/hplsql/Hplsql.g4, our syntax is the same as hplsql
image

such as:
`DROP PROCEDURE (IF EXISTS)? name=multipartIdentifier`
or
`DROP (PROCEDURE | PROC) (IF EXISTS)? name=multipartIdentifier`

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/PLParser.g4 (hunk 5 above)
Comment (xinyiZzz) [reply fe/fe-core/src/main/antlr4/org/apache/doris/nereids/PLParser.g4:328]: sure, my WeChat id is CatCatBlueBlue.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/DropProcedureCommand.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        this.procedureName = Objects.requireNonNull(procedureName, "procedureName is null");
+        this.source = Objects.requireNonNull(source, "source is null");
+    }
+
+    @Override
+    public void run(ConnectContext ctx, StmtExecutor executor) throws Exception {
+        client.dropPlsqlStoredProcedure(procedureName.getName(), procedureName.getCtl(), procedureName.getDb());
+        // TODO, removeCached needs to be synchronized to all Observer FEs.
+        // Even if it is always executed on the Master FE, it still has to deal with Master switching.
+        ctx.getPlSqlOperation().getExec().functions.removeCached(procedureName.toString());
```
Comment (xinyiZzz) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/DropProcedureCommand.java:58]:
```
client.dropPlsqlStoredProcedure(procedureName.getName(), procedureName.getCtl(), procedureName.getDb());
        // TODO, removeCached needs to be synchronized to all Observer FEs.
        // Even if it is always executed on the Master FE, it still has to deal with Master switching.
        ctx.getPlSqlOperation().getExec().functions.removeCached(procedureName.toString());
```
I think above code can be replaced with:
```
ctx.getPlSqlOperation().getExec().functions.remove(procedureName);
```

In fact `CreateProcedureCommand` also requires similar modifications. I missed it in my last PR.

**Comment by xinyiZzz** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/DropProcedureCommand.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    public DropProcedureCommand(FuncNameInfo procedureName, String source) {
+        super(PlanType.DROP_PROCEDURE_COMMAND);
+        this.client = new PlsqlMetaClient();
+        this.procedureName = Objects.requireNonNull(procedureName, "procedureName is null");
+        this.source = Objects.requireNonNull(source, "source is null");
+    }
+
+    @Override
+    public void run(ConnectContext ctx, StmtExecutor executor) throws Exception {
+        client.dropPlsqlStoredProcedure(procedureName.getName(), procedureName.getCtl(), procedureName.getDb());
```
Comment (xinyiZzz) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/DropProcedureCommand.java:55]: `DorisFunctionRegistry.remove()` contain this line, can delete it

**Review by xinyiZzz — APPROVED** [review]
LGTM

**Comment by xinyiZzz** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]

**Pull request merged**